### ISSUE TITLE
Allow overrides based on function call

### ISSUE THREAD

**Issue by overlookmotel** [eslint/eslint#14727]
**The version of ESLint you are using.**

7.29.0

**The problem you want to solve.**

In some cases a glob pattern is not sufficient to configure an override.

For example:

NodeJS allows an application to contain a mix of CommonJS and ESM JS files. ESM files can be differentiated from CommonJS by (1) use of the `.mjs` extension or (2) by using the `type` field in `package.json`.

I would like to set up a boilerplate ESLint shared config which correctly interprets JS files as CommonJS or ESM as per NodeJS's rules.

Dealing with the `.mjs` case is easy:

```js
{
  parserOptions: {
    sourceType: 'script'
  },
  overrides: [ {
    files: ['*.mjs'],
    parserOptions: {
      sourceType: 'module'
    }
  } ]
}
```

However, the case where `sourceType` for a `.js` file depends on a `package.json` file somewhere above it in the directory hierarchy, I cannot find a way to configure ESLint to handle this.

I imagine there are other use cases where how to parse/lint a file is dependent on more than the file path alone.

**Your take on the correct solution to problem.**

One solution would be if `overrides` could be configured with a function. e.g.:

```js
const getPackageType = require('get-package-type').sync;

module.exports = {
  parserOptions: {
    sourceType: 'script'
  },
  overrides: [ {
    files: ( path ) => {
      if ( /.cjs$/.test(path) ) return false;
      if ( /.mjs$/.test(path) ) return true;
      return getPackageType(path) === 'module';
    },
    parserOptions: {
      sourceType: 'module'
    }
  } ]
};
```

If it's undesirable to re-use the `files` property for this, could alternatively use another property called e.g. `test`.

Either way, I don't think this would constitute a breaking change.

**Are you willing to submit a pull request to implement this change?**

I sure am!

**Comment by btmills** [comment]
We’re working on a new config system, so we’re not likely to add any new features to the existing system. There are a few other ways you could tackle this particular problem within the existing config system. In increasing order of magic:

1. Manually add overrides for directories containing packages with the module type field.
2. In those `package.JSON` files, use the [`eslintConfig` key](https://eslint.org/docs/user-guide/configuring/configuration-files) to specify `”sourceType”: “module”`.
3. In `.eslintrc.JS` find ESM packages and generate the override file pattern array at runtime.

**Comment by nzakas** [comment]
We aren’t moving forward with this, so closing.
